### PR TITLE
Fix crash when direct boot

### PR DESCRIPTION
**Pull request by aleksandr-dorofeev**
* [x] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?
* ~[ ] Did you check UI tests on the sample app? They are not executed on CI.~

## Description

Added protected storage for data in direct boot mode.

## Related PRs or issues

#1599
[AB#91315](https://msmobilecenter.visualstudio.com/Mobile-Center/_workitems/edit/91315)

### CODE REVIEW

**Review by MatkovIvan — CHANGES_REQUESTED** [review]

**Comment by DmitriyKirakosyan** [comment]
@aleksandr-dorofeev  I think we should also modify the existing test(s) of AppCenter.run and check that we use contact from `ApplicationContextUtils.getApplicationContext`.

**Comment by aleksandr-dorofeev** [comment]
> @aleksandr-dorofeev I think we should also modify the existing test(s) of AppCenter.run and check that we use context from `ApplicationContextUtils.getApplicationContext`.

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/AbstractAppCenterTest.java [hunk 1]
```diff
@@ -71,7 +71,8 @@
         ShutdownHelper.class,
         InstrumentationRegistryHelper.class,
         NetworkStateHelper.class,
-        JSONUtils.class
+        JSONUtils.class,
+        ApplicationContextUtils.class
```
Comment (MatkovIvan) [inline sdk/appcenter/src/test/java/com/microsoft/appcenter/AbstractAppCenterTest.java:75]: Please sort this list alphabetically

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/AbstractAppCenterTest.java [hunk 2]
```diff
@@ -124,6 +128,10 @@ public void setUp() throws Exception {
         mApplicationInfo.flags = ApplicationInfo.FLAG_DEBUGGABLE;
         when(mApplication.getApplicationInfo()).thenReturn(mApplicationInfo);
 
+        /* Mock ApplicationContextUtils. */
+        mockStatic(ApplicationContextUtils.class);
+        when(ApplicationContextUtils.getApplicationContext(mApplication)).thenReturn(mContext);
+
```
Comment (MatkovIvan) [inline sdk/appcenter/src/test/java/com/microsoft/appcenter/AbstractAppCenterTest.java:134]: No blank line without comment

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java [hunk 3]
```diff
@@ -190,8 +190,8 @@ public void startTargetTokenThenStartWithAppSecretTest() {
         DummyService service = DummyService.getInstance();
         assertTrue(AppCenter.getInstance().getServices().contains(service));
         verify(service).getLogFactories();
-        verify(service).onStarted(any(Context.class), any(Channel.class), isNull(), eq(DUMMY_TRANSMISSION_TARGET_TOKEN), eq(true));
-        verify(service, never()).onStarted(any(Context.class), any(Channel.class), eq(DUMMY_APP_SECRET), isNull(), eq(true));
+        verify(service).onStarted(eq(mContext), any(Channel.class), isNull(), eq(DUMMY_TRANSMISSION_TARGET_TOKEN), eq(true));
+        verify(service, never()).onStarted(eq(mContext), any(Channel.class), eq(DUMMY_APP_SECRET), isNull(), eq(true));
```
Comment (MatkovIvan) [inline sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java:194]: Same

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java [hunk 4]
```diff
@@ -231,8 +231,8 @@ public void configureTwiceTest() {
         DummyService service = DummyService.getInstance();
         assertTrue(AppCenter.getInstance().getServices().contains(service));
         verify(service).getLogFactories();
-        verify(service).onStarted(any(Context.class), any(Channel.class), eq(DUMMY_APP_SECRET), isNull(), eq(true));
-        verify(service, never()).onStarted(any(Context.class), any(Channel.class), eq(DUMMY_APP_SECRET + "a"), isNull(), eq(true));
+        verify(service).onStarted(eq(mContext), any(Channel.class), eq(DUMMY_APP_SECRET), isNull(), eq(true));
+        verify(service, never()).onStarted(eq(mContext), any(Channel.class), eq(DUMMY_APP_SECRET + "a"), isNull(), eq(true));
```
Comment (MatkovIvan) [inline sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java:235]: Same

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java [hunk 5]
```diff
@@ -682,101 +682,101 @@ private void checkStartedWithoutAppSecret() {
...
         /* Verify start not called again (1 total call). */
-        verify(DummyService.getInstance()).onStarted(any(Context.class), any(Channel.class), any(), any(), anyBoolean());
+        verify(DummyService.getInstance()).onStarted(eq(mContext), any(Channel.class), any(), any(), anyBoolean());
 
         /* And not updated either. */
         verify(DummyService.getInstance(), never()).onConfigurationUpdated(anyString(), anyString());
 
         /* Verify the second service was not started as was part of second secret configuration. */
-        verify(AnotherDummyService.getInstance(), never()).onStarted(any(Context.class), any(Channel.class), anyString(), anyString(), anyBoolean());
+        verify(AnotherDummyService.getInstance(), never()).onStarted(eq(mContext), any(Channel.class), anyString(), anyString(), anyBoolean());
```
Comment (MatkovIvan) [inline sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java:699]: Same

**Review by MatkovIvan — CHANGES_REQUESTED** [review]

**Review by AlexeyBudynkov — APPROVED** [review]

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/ApplicationContextUtilsTest.java [hunk 6]
```diff
@@ -104,6 +105,29 @@ public void getContextWhenVersionIsLowerThanN() {
...
     }
 
+    @Test
+    public void ifTheDeviceAPIVersionSupportsDeviceProtectedStorage() {
+
+        /* Mock SDK_INT to N. */
+        Whitebox.setInternalState(Build.VERSION.class, "SDK_INT", Build.VERSION_CODES.N);
+
+        /* Method call. */
+        ApplicationContextUtils.isDeviceProtectedStorage(mProtectedContext);
```
Comment (MatkovIvan) [inline sdk/appcenter/src/test/java/com/microsoft/appcenter/ApplicationContextUtilsTest.java:115]: Please add return value check

**Review by MatkovIvan — APPROVED** [review]

**Review by DmitriyKirakosyan — APPROVED** [review]

**Pull request merged**